book/src/error_handling.md

```diff
@@ -31,27 +49,116 @@ pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_divide(
         result
     } else {
         // Divisor is zero -- throw an exception!
-        env.throw_new("java/lang/ArithmeticException", "Attempting to divide by zero.");
-        // It doesn't matter what we return now, since the JNI will see that an
-        // exception was thrown.
+        env.throw_new("java/lang/Exception", "Attempting to divide by zero.");
+        // A dummy value must be returned, since the native method retains control
+        // even after throwing an Exception.
         0
     }
 }
 ```
+    
+## Wrapping Result Handling into a Function
+
+It can be helpful to wrap common error handling into one function, so that there
+is less repeated JNI code. For ease of implementation, we'll use
+`anyhow::Error`, and make the caller pass the `dummy_value`, which is returned
+only on error paths. The `JNIEnv` APIs all throw `Result<T,E>`, and the
+`JavaException` variant marks that an exception has already been thrown - so for
+that variant alone, we will skip throwing an Exception. You may come up with
+your own Error hierarchies, which could have a different logic.
+
+```rust
+fn try_java<F, T>(env: JNIEnv, dummy_value: T, f: F) -> T
+where
+    F: FnOnce() -> Result<T, anyhow::Error>,
+{
+    match f() {
+        Ok(s) => s,
+        Err(e) => {
+            if let Some(e) = e.downcast_ref::<jni::errors::Error>() {
+                match e {
+                    // Since JavaException implies an Exception has already been thrown, 
+                    // don't throw another one.
+                    jni::errors::Error::JavaException => {},
+                    _ => {
+                        let _ = env.throw_new("java/lang/Exception", e.to_string());
+                    }
+                }
+            } else {
+            // This branch handles all other error types, such as those returned by
+            // all non-JNI code
+                let _ = env.throw_new("java/lang/Exception", e.to_string());
```

Comment:
Since it appears as a library method, using _checked_ "java/lang/Exception" is not generally applicable (as it is valid to have native method declarations without checked exceptions). RuntimeException seems more appropriate.

Reply:
I wonder if there might be client code that returns non-jni error, but has a pending Java exception (e.g., by ignoring a jni error that they got; and getting an error from somewhere else).

Reply:
Thanks :)

Right, `RuntimeException` is much better. A colleague also pointed out that we'd need to decorate the native functions with `throws Exception`, which wasn't done in the examples. I'll update this to `RuntimeException`

I agree, this doesn't handling pending exceptions properly. Any unchecked (doesn't check for an exception, when an exception is possible) JNI method can lead to problems. I'll review exonum and see if we can get a better pattern here.

Reply:
> Any unchecked (doesn't check for an exception, when an exception is possible) JNI method can lead to problems

I think all methods that are required by the spec do check for exceptions (if not — that's a bug). I mostly though of some strange client code that, for example, throws an exceptions/invokes a method that throws an exception, and does not return that error straight away. Hopefully, you can just pull this check a little higher to cover any such cases. On top of that, if an actual error does not match JniError:Exception when there is a pending excpetion in Java, we can log such a discrepancy.

Reply:
Sounds good. I think the strange client code is likely common, so hopefully we can figure something out here. I still need to experiment with the jni-rs APIs to understand a little bit better.

One case I'm thinking about is dependent on jni-rs APIs. If call_method calls some Java method that throws an Exception, would the result be `Err` or `Ok`? Intuitively, I'd expect it to be `Ok`, and so some additional checks are necessary that the writing doesn't advertise.

Reply:
Perhaps this is an example?

```
try_java(|| {
  env.call_method(obj, "throwAnException", "()V"); // ignore exception.
  env.call_method(obj, "doSomethingElse", "()V"); // exception still pending
});
```

Reply:
Maybe something like this instead?
```rust

fn try_java<F, T>(env: JNIEnv, error_value: T, f: F) -> T
    where
        F: FnOnce() -> Result<T, anyhow::Error>,
{
    match f() {
        Ok(s) => s,
        Err(e) => {
            // Only throw an exception is one isn't already pending.
            if !env.exception_check().unwrap() {
                env.throw_new("java/lang/RuntimeException", e.to_string())
            }
            error_value
        }
    }
}
```

Reply:
> So far, I haven't found a JNI function call that returns Ok when it finishes with an exception pending

There might be some examples that may return Error, but of a different type (e.g., if they check an int error code first). But generally, methods that wrap JNI APIs that can throw exceptions use a macro that checks them for you (so that the client code rarely, if ever, have to). For example:
https://github.com/jni-rs/jni-rs/blob/master/src/wrapper/macros.rs#L21

> I think this is probably OK to start with.

+1; exception_check must not be particularly expensive: https://github.com/jni-rs/jni-rs/blob/master/benches/api_calls.rs#L206-L217